### PR TITLE
De-redundantize Error kinds.

### PR DESCRIPTION
**Pull request by jsha**
Change "Bad" to "Invalid" in error names, mimicking io::Error::ErrorKind.

Change InvalidProxyCreds to ProxyUnauthorized.

Change DnsFailed to just Dns (the fact that there was a failure is implicit
in the fact that this was an error).

### CODE REVIEW

File: src/error.rs [hunk 1]
```diff
@@ -99,29 +99,27 @@ impl Error {
 #[derive(Debug, PartialEq, Clone, Copy)]
 pub enum ErrorKind {
     /// The url could not be understood.
-    BadUrl,
+    Url,
```
Comment (algesten) [inline src/error.rs:102]: InvalidUrl?

**Review by algesten — COMMENTED** [review]
In two minds about this.

I think the archetype is `std::io::ErrorKind`, which very clearly has stuff like `InvalidInput`, `InvalidData`.

"Bad" might not be a good word. "InvalidUrl" seems ok.

**Comment by jsha** [comment]
Good point about io::Error having Invalid. I've changed to InvalidUrl. Are there any others that you think should have `Invalid`? E.g. InvalidStatus or InvalidHeader?

**Comment by jyn514** [comment]
'InvalidStatus' seems to mean the status is itself invalid, like a status of 0 or something. I think 'ErrorStatus' or something would be more descriptive.

**Comment by jsha** [comment]
> 'InvalidStatus' seems to mean the status is itself invalid, like a status of 0 or something. I think 'ErrorStatus' or something would be more descriptive.

Actually, the ErrorKind `InvalidStatus` (or `Status` in this PR) reflects a failure to parse the status line - for instance a status of 0. For this PR, I'm going to revert it to `BadStatus` just so we don't have lots of conflicts with #266. Later, though, we should change this to merge BadStatus and BadHeader into just one ErrorKind, `InvalidHeader`. That avoids any potential confusion with `error::Error::Status(u16, Response)`.

**Review by algesten — APPROVED** [review]
Let's do it!

**Pull request merged**